### PR TITLE
Fixing incorrect default TLS min and TLS max versions

### PR DESCRIPTION
**Pull request by bushwhackr**
**Description:** Fixing a bug, default TLS version and description does not match

**Link to tracking Issue:** #5430

### CODE REVIEW

**Comment by bushwhackr** [comment]
wondering if just using the crypto/tls defaults of TLS 1.2 as client and TLS 1.0 as server would be sufficient. 

**Default in crypto/tls**
https://github.com/golang/go/blob/47f806ce81aac555946144f112b9f8733e2ed871/src/crypto/tls/common.go#L684

**crypto/tls function that sets the allowed TLS versions**
https://github.com/golang/go/blob/47f806ce81aac555946144f112b9f8733e2ed871/src/crypto/tls/common.go#L984

**Review by bogdandrutu — COMMENTED** [review]
Can we add tests for this?

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Also, you are not fixing the description, you are fixing code, so please modify the title accordingly.

File: config/configtls/configtls.go
Comment (bogdandrutu) [reply config/configtls/configtls.go:195]: Is it guaranteed that the TLS version representation as a number is sorted?

File: config/configtls/configtls.go
Comment (tigrannajaryan) [reply config/configtls/configtls.go:186]: @bogdandrutu your analysis look correct to me. So, essentially you are saying that we want:
- On clients if MinVersion is missing then set it to 1.2.
- On servers if MinVersion is missing then set it to 1.0.
- Both on clients and servers if MaxVersion is missing then set it to 1.3.

File: config/configtls/configtls.go
Comment (bushwhackr) [reply config/configtls/configtls.go:195]: I referred to these lines which they must be inherently sorted
https://github.com/golang/go/blob/master/src/crypto/tls/common.go#L984
https://github.com/golang/go/blob/master/src/crypto/tls/common.go#L1003

File: config/configtls/configtls.go
Comment (tigrannajaryan) [reply config/configtls/configtls.go:186]: > Not setting a default value at all crypto/tls will handle the TLS minVersion and maxVersions for us. Should I proceed with updating the PR?

That's a good point. Just pass the zero values and let crypto/tl handle the defaults.

File: config/configtls/configtls.go
Comment (bushwhackr) [reply config/configtls/configtls.go]: Removed, let crypto/tls throw the error if minTLS > maxTLS

File: config/configtls/configtls.go
Comment (bushwhackr) [reply config/configtls/configtls.go]: I've updated the PR so that we do not add a default via opentelemetry, crypto/tls will handle the default TLS versions.

File: config/configtls/configtls.go
Comment (codeboten) [inline config/configtls/configtls.go:45]: I'd remove the last line here, otherwise this will eventually get out of sync if/when crypto/tls changes its defaults

**Review by codeboten — COMMENTED** [review]
Just one comment

File: config/configtls/configtls.go
Comment (bushwhackr) [reply config/configtls/configtls.go]: Altered

File: config/configtls/configtls.go
Comment (bogdandrutu) [reply config/configtls/configtls.go]: Have we looked into this https://github.com/open-telemetry/opentelemetry-collector/pull/4503 and the CWE-327? Do we want to enforce on the server side "1.2" as well?

Comment (bushwhackr) [reply config/configtls/configtls.go]: Was thinking about this problem and I agree that, TLS 1.0 and 1.1 has been deprecated.
https://datatracker.ietf.org/doc/html/rfc8996

However, golang tls package picks the most secure cipher suite for client negotiation; unless the server has **explicitly** set the `MaxVersion` < "1.2" a TLS cipher of TLS1.2 and above will be used for negotiation. 
https://github.com/golang/go/blob/55590f3a2b89f001bcadf0df6eb2dde62618302b/src/crypto/tls/handshake_client.go#L39

For server side, the order of cipher is decided by the client via their preference order. And in this regard, there is a chance that the client has specified a insecure TLS version to be used. https://github.com/golang/go/blob/55590f3a2b89f001bcadf0df6eb2dde62618302b/src/crypto/tls/handshake_server.go#L159.
My opinion is that clients are incharge of the level of secrecy of the data that is being sent. If they are using a client that does not support higher TLS version it is up to them to upgrade it. Not the server's duty to force secrecy but rather be as compatible as possible.

That being said, the above is all my opinion and I'm open in changing it.

Comment (bogdandrutu) [reply config/configtls/configtls.go]: @tigrannajaryan @codeboten please tell me if it is fine to downgrade the server min version.

Comment (tigrannajaryan) [reply config/configtls/configtls.go]: My understand is that this PR is about fixing the defaults to match the default of what the Go's standard library defines, so let's keep it that way. The code in this PR looks correct to me as it is.

If we want to require higher version number than Go's standard library then it is best to discuss it in in a separate issue/PR.

Comment (bogdandrutu) [reply config/configtls/configtls.go]: This change just delegates to the golang compiler/vm to pick the min/max TLS. Am I wrong? We are already building the binaries with golang 1.8 (see https://github.com/open-telemetry/opentelemetry-collector-releases/blob/main/.github/workflows/ci-goreleaser.yaml#L35), but we have "go 1.17" in the go.mod. Does that mean that the min version is 1.0?

Comment (mx-psi) [reply config/configtls/configtls.go]: Oh, I didn't realize we were already using Go 1.18 on the official binaries. Then I believe regardless of what `go.mod` says, this will not mean a change on the official binaries, so I feel like doing this is okay

Comment (bushwhackr) [reply config/configtls/configtls.go]: Just rebased the PR to `main`, is there anything blocking this PR now? I'll be happy to follow through with the changes.

**Review by mx-psi — COMMENTED** [review]
LGTM, but there is one unrelated change. Could you move it to a different PR?

File: docs/release.md
Comment (bushwhackr) [reply docs/release.md:10]: will do

**Review by mx-psi — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Comment by codeboten** [comment]
@tigrannajaryan @bogdandrutu please review and approve the PR if your concerns have been addressed.

**Review by bogdandrutu — APPROVED** [review]

File: config/configtls/README.md
Comment (jpkrohling) [inline config/configtls/README.md:39]: As I understood it, 1.0 is used as the default version on the server-side. This is *not* what we want here, on a newer server/library: the recommendation for a few years has been already to default to 1.2 as the minimum version, allowing users to use 1.0 or SSLvX only when absolutely needed.

**Pull request merged**
